resolver/parallel_best_resolver.go

```diff
@@ -153,88 +137,137 @@ func (r *ParallelBestResolver) Name() string {
+	result := make([]string, len(r.resolvers))
+	for i, s := range r.resolvers {
+		result[i] = fmt.Sprintf("%s", s.resolver)
 	}
 
-	return fmt.Sprintf("parallel upstreams '%s'", strings.Join(result, "; "))
+	return fmt.Sprintf("%s (resolverCount: %d, retryWithDifferentResolver: %t) upstreams '%s (%s)'",
+		parallelResolverType, r.resolverCount, r.retryWithDifferentResolver, r.groupName, strings.Join(result, ","))
 }
 
 // Resolve sends the query request to multiple upstream resolvers and returns the fastest result
 func (r *ParallelBestResolver) Resolve(request *model.Request) (*model.Response, error) {
 	logger := log.WithPrefix(request.Log, parallelResolverType)
 
-	var resolvers []*upstreamResolverStatus
-	for _, r := range r.resolversPerClient {
-		resolvers = r
+	if len(r.resolvers) == 1 {
+		logger.WithField("resolver", r.resolvers[0].resolver).Debug("delegating to resolver")
 
-		break
+		return r.resolvers[0].resolver.Resolve(request)
 	}
 
-	if len(resolvers) == 1 {
-		logger.WithField("resolver", resolvers[0].resolver).Debug("delegating to resolver")
+	ctx := context.Background()
 
-		return resolvers[0].resolver.Resolve(request)
+	// using context with timeout for random upstream strategy
+	if r.resolverCount == 1 {
+		var cancel context.CancelFunc
+
+		logger = log.WithPrefix(logger, "random")
+		timeout := config.GetConfig().Upstreams.Timeout
+
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(timeout))
+		defer cancel()
 	}
 
-	r1, r2 := pickRandom(resolvers)
-	logger.Debugf("using %s and %s as resolver", r1.resolver, r2.resolver)
+	resolvers := pickRandom(r.resolvers, r.resolverCount)
+	ch := make(chan requestResponse, len(resolvers))
 
-	ch := make(chan requestResponse, resolverCount)
+	for _, resolver := range resolvers {
+		logger.WithField("resolver", resolver.resolver).Debug("delegating to resolver")
 
-	var collectedErrors []error
+		go resolver.resolve(request, ch)
+	}
 
-	logger.WithField("resolver", r1.resolver).Debug("delegating to resolver")
+	response, collectedErrors := evaluateResponses(ctx, logger, ch, resolvers)
+	if response != nil {
+		return response, nil
+	}
 
-	go r1.resolve(request, ch)
+	if !r.retryWithDifferentResolver {
+		return nil, fmt.Errorf("resolution failed: %w", errors.Join(collectedErrors...))
+	}
+
+	return r.retryWithDifferent(logger, request, resolvers)
+}
+
+func evaluateResponses(
+	ctx context.Context, logger *logrus.Entry, ch chan requestResponse, resolvers []*upstreamResolverStatus,
+) (*model.Response, []error) {
+	collectedErrors := make([]error, 0, len(resolvers))
+
+	for len(collectedErrors) < len(resolvers) {
+		select {
+		case <-ctx.Done():
+			// this context currently only has a deadline when resolverCount == 1
+			field := logrus.Fields{"resolver": resolvers[0].resolver}
+			if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+				logger.WithFields(field).Debug("upstream exceeded timeout, trying other upstream")
```

Comment:
Noticed we can simplify the field passing setup:
```suggestion
			// this context currently only has a deadline when resolverCount == 1
			if errors.Is(ctx.Err(), context.DeadlineExceeded) {
				logger.
					WithField("resolver", resolvers[0].resolver).
					Debug("upstream exceeded timeout, trying other upstream")
```